finality-aleph/src/new_network/connection_manager/session.rs

```diff
@@ -151,16 +178,19 @@ impl Handler {
     /// If successful returns a set of addresses that we should be connected to.
     pub async fn update(
         &mut self,
-        keychain: KeyBox,
+        authority_index_and_pen: Option<(NodeIndex, AuthorityPen)>,
+        authority_verifier: AuthorityVerifier,
         addresses: Vec<Multiaddr>,
     ) -> Result<Vec<Multiaddr>, AddressError> {
-        let (own_authentication, own_peer_id) =
-            construct_authentication(&keychain, self.session_id(), addresses).await?;
+        let (session_info, own_peer_id) =
+            construct_authentication(&authority_index_and_pen, self.session_id(), addresses)
+                .await?;
         let authentications = self.authentications.clone();
         self.authentications = HashMap::new();
         self.peers_by_node = HashMap::new();
-        self.keychain = keychain;
-        self.own_authentication = own_authentication;
+        self.authority_index_and_pen = authority_index_and_pen;
+        self.authority_verifier = authority_verifier;
+        self.session_info = session_info;
         self.own_peer_id = own_peer_id;
         for (_, (auth, maybe_auth)) in authentications {
```

Comment:
I think this is more a comment for me (since I designed this originally) so I'll try explaining myself.

First of all, yeah, there is definitely annoying duplication here, and I would be quite happy if it could get deduplicated.

For why this is coded this way despite the above let me describe the usual path a handler takes. It gets created whenever we start caring about a session (usually a couple sessions ahead). Sometimes some of the session keys of the validators can change after creation. When this happens we would like to be able to stay connected to all the validators for which the authentications we have are still valid (often this would be almost all of them). This is why we want to recheck all the authentications we already have. It is also likely that we get an authentication signed with new keys before we are aware they are changing – that's why we keep one additional authentication and also recheck it here.

There are ways of rewriting the code keeping the above use pattern simple, most obviously by rewriting `update` to take `self` by value, but I'm not convinced they are actually better in this case. Some concrete options you think are better would be useful.

Reply:
```suggestion
    pub async fn update(
        &mut self,
        authority_index_and_pen: Option<(NodeIndex, AuthorityPen)>,
        authority_verifier: AuthorityVerifier,
        addresses: Vec<Multiaddr>,
    ) -> Result<Vec<Multiaddr>, AddressError> {
        let authentications = std::mem::take(&mut self.authentications);
        *self = Handler::new(
            authority_index_and_pen,
            authority_verifier,
            self.session_id(),
            addresses,
        )
        .await?;
        for (_, (auth, maybe_auth)) in authentications {
            print!(
                "normal authentication: {:?}",
                self.handle_authentication(auth)
            );
            if let Some(auth) = maybe_auth {
                print!(
                    "alternative authentication: {:?}",
                    self.handle_authentication(auth)
                );
            }
        }
        Ok(self
            .authentications
            .values()
            .flat_map(|((auth_data, _), _)| auth_data.addresses.iter().cloned())
            .collect())
    }
```

Reply:
Huh, cool.

Reply:
Done with small change (without `mem::take` call)